torch/onnx/__init__.py

```diff
@@ -139,23 +142,51 @@ def _run_symbolic_method(op_name, symbolic_fn, args):
         raise
 
 
-def _add_attribute(node, key, value):
-    """ initializes the right attribute based on type of value """
-    m = attr_pattern.match(key)
-    if m is None:
-        raise IndexError((
-            "Invalid attribute specifier '{}' names " +
-            " must be suffixed with type, e.g. 'dim_i' or 'dims_i'").format(key))
-    name, kind = m.group(1), m.group(2)
+def _is_onnx_list(value):
     if not isinstance(value, string_classes) and not torch.is_tensor(value) and isinstance(value, collections.Iterable):
-        kind += "s"
-    return getattr(node, kind + '_')(name, value)
+        return True
+    return False
+
+
+def _add_attribute(node, key, value, aten):
+    """ initializes the right attribute based on type of value """
+    if aten:
+        name = key
+
+        def infer_kind(value):
```

Comment:
We should not need the `infer_kind` logic at all in the ATen case. Think about where the attributes come from: you pulled them out of a `Node` using `attrs = {k: n[k] for k in n.attributeNames()}`. The `n[k]` drops precisely the type information you actually want to use to determine the ATen types. You want the `kindOf` method: `n.kindOf("key")` will give you `f` for float, `fs` for list of floats, etc. Then all of this code can go away.

In this regime, you should also delete the `aten` kwarg from `_graph_at`, and just make a new function which does this construction. These two functions aren't doing the same thing at all!

Reply:
I see. Will call `kindOf`.